File: Nutrisolutions-FRONT/src/app/services/nutritionists.service.ts
import { inject, Injectable } from '@angular/core';
import { NutritionistModel, StatusEnum } from '../models/nutritionist.model';
// import { generateFakeNutritionist } from '../core/helpers/faker.helper';
import { APP_API } from '../core/constants/constants.config';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ClientModel } from '../models/client.model';
import { ExperienceEnum } from '../models/recipe.model';

@Injectable({
  providedIn: 'root',
})
export class NutritionistsService {
  private nutritionists: NutritionistModel[] = [];

  #nutritionists: NutritionistModel[] = [];

  apiUrl = APP_API.base_url + '/nutritionists';

  constructor() {}
  http = inject(HttpClient);

  getAllNutritionists(
    page: number = 1,
    limit: number = 12,
    searchText?: string,
    experience?: ExperienceEnum,
    status?: StatusEnum
  ): Observable<{
    data: NutritionistModel[];
    total: number;
  }> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('limit', limit.toString());
    if (searchText?.trim()) {
      params = params.set('searchText', searchText.trim());
    }
    if (experience) {
      params = params.set('experience', experience);
    }
    if (status) {
      params = params.set('status', status);
    }
    return this.http.get<{
      data: NutritionistModel[];
      total: number;
    }>(this.apiUrl, { params });
  }
  getNutritionistsCount(): Observable<{ total: number }> {
    return this.http.get<{ total: number }>(this.apiUrl + '/count');
  }
  getBestNutritionists(): Observable<NutritionistModel[]> {
    return this.http.get<NutritionistModel[]>(this.apiUrl + '/top');
  }
  getNutritionistById(id: string): Observable<NutritionistModel> {
    return this.http.get<NutritionistModel>(`${this.apiUrl}/${id}`);
  }
  getPatientsByNutritionist(id: string): Observable<ClientModel[]> {
    return this.http.get<ClientModel[]>(`${this.apiUrl}/${id}/patients`);
  }
  deleteNutritionist(id: string): Observable<{ count: number }> {
    return this.http.delete<{ count: number }>(`${this.apiUrl}/${id}`);
  }
  // addNutritionist(
  //   nutritionnist: NutritionistModel
  // ): Observable<NutritionistModel> {
  //   return this.http.post<NutritionistModel>(this.apiUrl, nutritionnist);
  // }
  updateNutritionist(
    id: string,
    status: StatusEnum
  ): Observable<NutritionistModel> {
    return this.http.patch<NutritionistModel>(`${this.apiUrl}/${id}`, {
      status,
    });
  }

  // getAllNutritionists(): NutritionistModel[] {
  //   return this.nutritionists.length
  //     ? this.nutritionists
  //     : this.generateFakeNutritionistList(15);
  // }
  // // Get all nutritionists
  // getBestNutritionists(): NutritionistModel[] {
  //   return this.generateFakeNutritionistList(4);
  // }

  // generateFakeNutritionistList(count: number): NutritionistModel[] {
  //   return Array.from({ length: count }, () => generateFakeNutritionist());
  // }

  // // Add a new nutritionist
  // addNutritionist(nutritionist: NutritionistModel): void {
  //   this.nutritionists.push(nutritionist);
  // }

  // // Find nutritionist by ID
  // getNutritionistById(id: string): NutritionistModel | undefined {
  //   return this.nutritionists.find((nutritionist) => nutritionist.id === id);
  // }

  // // Update a nutritionist
  // updateNutritionist(
  //   id: string,
  //   updatedData: Partial<NutritionistModel>
  // ): void {
  //   const index = this.nutritionists.findIndex(
  //     (nutritionist) => nutritionist.id === id
  //   );
  //   if (index > -1) {
  //     this.nutritionists[index] = {
  //       ...this.nutritionists[index],
  //       ...updatedData,
  //     };
  //   }
  // }

  // // Delete a nutritionist
  // deleteNutritionist(id: string): void {
  //   this.nutritionists = this.nutritionists.filter(
  //     (nutritionist) => nutritionist.id !== id
  //   );
  // }
}

File: nutrisolutions_mobile/lib/data/services/nutritionists_service.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'package:dio/dio.dart';

import 'package:nutrisolutions_mobile/core/constants/app_constants.dart';
import 'package:nutrisolutions_mobile/data/services/dio_client.dart';

import '../models/nutritionist_model.dart';

class NutritionistsResponse {
  List<NutritionistModel> nutritionists;
  int totalCount;
  NutritionistsResponse({
    required this.nutritionists,
    required this.totalCount,
  });
}

class NutritionistsService {
  static const String apiUrl = '${AppApi.baseUrl}/nutritionists';

  final Dio _dio;

  NutritionistsService() : _dio = DioClient(baseUrl: apiUrl).dio;

  Future<NutritionistsResponse> getAllNutritionists({
    int page = 1,
    int limit = 12,
    String? searchText,
    String? experience,
  }) async {
    try {
      final queryParameters = {
        'page': page,
        'limit': limit,
        if (searchText != null && searchText.trim().isNotEmpty)
          'searchText': searchText.trim(),
        if (experience != null) 'experience': experience,
      };

      final response = await _dio.get('', queryParameters: queryParameters);

      final List<dynamic> data = response.data['data'];
      final int totalNutritionistsCount = response.data['total'];
      print(data);
      final nutritionists =
          data.map((json) => NutritionistModel.fromMap(json)).toList();
      print('from nutritionists service: $nutritionists');
      return NutritionistsResponse(
          nutritionists: nutritionists, totalCount: totalNutritionistsCount);
    } catch (e) {
      throw Exception('Failed to load nutritionists: $e');
    }
  }

  Future<int> getNutritionistsCount() async {
    try {
      final response = await _dio.get('/count');
      return response.data['total'];
    } catch (e) {
      throw Exception('Failed to fetch nutritionists count: $e');
    }
  }

  Future<NutritionistModel> getNutritionistById(String id) async {
    try {
      final response = await _dio.get('/$id');
      return NutritionistModel.fromMap(response.data);
    } catch (e) {
      throw Exception('Failed to fetch nutritionist: $e');
    }
  }

  Future<List<NutritionistModel>> getBestNutritionists() async {
    try {
      final response = await _dio.get('/top');
      final List<dynamic> data = response.data;

      return data.map((json) => NutritionistModel.fromMap(json)).toList();
    } catch (e) {
      throw Exception('Failed to fetch top nutritionists: $e');
    }
  }

  Future<bool> deleteNutritionist(String id) async {
    try {
      final response = await _dio.delete('/$id');
      return response.data['count'] > 0;
    } catch (e) {
      throw Exception('Failed to delete nutritionist: $e');
    }
  }

  Future<NutritionistModel> addNutritionist(
      NutritionistModel nutritionist) async {
    try {
      final response = await _dio.post(
        '',
        data: nutritionist.toJson(),
      );
      return NutritionistModel.fromJson(response.data);
    } catch (e) {
      throw Exception('Failed to add nutritionist: $e');
    }
  }

  Future<NutritionistModel> updateNutritionist(
      String id, NutritionistModel nutritionist) async {
    try {
      final response = await _dio.patch(
        '/$id',
        data: nutritionist.toJson(),
      );
      return NutritionistModel.fromJson(response.data);
    } catch (e) {
      throw Exception('Failed to update nutritionist: $e');
    }
  }
}
